### PR TITLE
Clarify mappings of SAM flags to read fields

### PR DESCRIPTION
**Pull request by dcolligan**
Based on discussion here https://github.com/ga4gh/server/pull/789

### CODE REVIEW

File: src/main/resources/avro/reads.avdl [hunk 1]
```diff
@@ -190,7 +190,12 @@ record ReadAlignment {
   /** The fragment is a PCR or optical duplicate (SAM flag 0x400). */
   union { null, boolean } duplicateFragment = null;
 
-  /** The number of reads in the fragment (extension to SAM flag 0x1) */
+  /**
+   The number of reads in the fragment.
+   A value of 1 means SAM flag 0x1 is unset.
+   A value of 2 means SAM flag 0x1 is set.
+   TODO what do other values mean, especially null?
```
Comment (macieksmuga) [inline src/main/resources/avro/reads.avdl]: can somebody familiar with reads comment on this? The null union here seems spurious to me based on the documentation so far.

**Comment by macieksmuga** [comment]
+1 modulo the question about null union in number of reads.
Since this PR is purely about improving the quality of documentation (not changing the meaning of any fields), I believe it's safe not to wait on this to be merged for the 0.6.0a1 release.

File: src/main/resources/avro/reads.avdl [hunk 2]
```diff
@@ -202,6 +207,12 @@ record ReadAlignment {
   The read ordinal in the fragment, 0-based and less than numberReads. This
   field replaces SAM flag 0x40 and 0x80 and is intended to more cleanly
   represent multiple reads per fragment.
+  A readNumber of 0 is equivalent to SAM flag 0x40 being set and 0x80 being
+  unset.
+  A readNumber of 1 is equivalent to SAM flag 0x80 being set and 0x40 being
+  unset.
+  A readNumber other than 0 or 1 means both SAM flags 0x40 and 0x80 are set.
```
Comment (calbach) [inline src/main/resources/avro/reads.avdl]: The wording throughout this PR seems off to me. SAM flags are not 'set' in the GA4GH APIs. Some property of the GA4GH API may be equivalent to a SAM flag, for instance our comment for `properPlacement`: `(equivalent to SAM flag 0x2)`.

File: src/main/resources/avro/reads.avdl [hunk 3]
```diff
@@ -190,7 +190,12 @@ record ReadAlignment {
   /** The fragment is a PCR or optical duplicate (SAM flag 0x400). */
   union { null, boolean } duplicateFragment = null;
 
-  /** The number of reads in the fragment (extension to SAM flag 0x1) */
+  /**
+   The number of reads in the fragment.
+   A value of 1 means SAM flag 0x1 is unset.
```
Comment (calbach) [inline src/main/resources/avro/reads.avdl]: Technically the mapping to SAM would be: 

```
1 -> 0x1 unset
[2, infinity) -> 0x1 set
```

I would guess that `null` means an unknown number of reads, or it shouldn't be allowed.

File: src/main/resources/avro/reads.avdl (hunk 2 above)
Comment (macieksmuga) [inline src/main/resources/avro/reads.avdl]: I _think_ that's what this PR is already saying. My naive understanding is that `readNumber` should correctly represent an arbitrary number of reads per fragment, as opposed to SAM's two-flag mechanism used to represent 0, 1 or many... (or should it be 1, 2 or many?)

Comment (richarddurbin) [inline src/main/resources/avro/reads.avdl]: I think that the current comment is clear and OK.  This is saying very clearly that each GA4GH API value of 
readNumber is equivalent to a particular combination of SAM flag states - there is no single equivalence to
any one tag.  It doesn’t matter whether flags are set or unset in the API - they are in SAM, and it is
their SAM state that is being referred to.
 
> On 16 Dec 2015, at 02:47, CH Albach <notifications@github.com> wrote:
> 
> In src/main/resources/avro/reads.avdl <https://github.com/ga4gh/schemas/pull/509#discussion_r47731145>:
> 
> > @@ -202,6 +207,12 @@ record ReadAlignment {
> >    The read ordinal in the fragment, 0-based and less than numberReads. This
> >    field replaces SAM flag 0x40 and 0x80 and is intended to more cleanly
> >    represent multiple reads per fragment.
> > +  A readNumber of 0 is equivalent to SAM flag 0x40 being set and 0x80 being
> > +  unset.
> > +  A readNumber of 1 is equivalent to SAM flag 0x80 being set and 0x40 being
> > +  unset.
> > +  A readNumber other than 0 or 1 means both SAM flags 0x40 and 0x80 are set.
> The wording throughout this PR seems off to me. SAM flags are not 'set' in the GA4GH APIs. Some property of the GA4GH API may be equivalent to a SAM flag, for instance our comment for properPlacement: (equivalent to SAM flag 0x2).
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ga4gh/schemas/pull/509/files#r47731145>.
> 




-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by david4096** [comment]
@dcolligan when you have a chance could you review, close, or update this?

**Comment by dcolligan** [comment]
@david4096 updated

File: src/main/proto/ga4gh/reads.proto [hunk 4]
```diff
@@ -151,13 +154,19 @@ message ReadAlignment {
...
+  // A read_number other than 0 or 1 means both SAM flags 0x40 and 0x80 are 
+  // set.
   int32 read_number = 8;
 
   // The read fails platform or vendor quality checks (SAM flag 0x200).
   bool failed_vendor_quality_checks = 9;
 
-  // The alignment for this alignment message. This field will be null if the
-  // read is unmapped.
+  // The alignment for this alignment message. This field will be zero if the
```
Comment (david4096) [inline src/main/proto/ga4gh/reads.proto]: `unset` maybe? does it come out 0?

**Comment by david4096** [comment]
Thanks @dcolligan there is still a lot of complexity hidden in the conversion from BAM to GA4GH read representation and giving implementors more to go on is helpful.

File: src/main/proto/ga4gh/reads.proto (hunk 4 above)
Comment (david4096) [inline src/main/proto/ga4gh/reads.proto]: Saying the field will be `0` confused me. I think you mean to say if it isn't set perhaps?

Comment (dcolligan) [inline src/main/proto/ga4gh/reads.proto]: yes, you're right... changing...

**Comment by david4096** [comment]
Thanks, reminds me that we need a way to get back only unmapped reads!

**Pull request closed without merging**